cpp/src/arrow/compute/kernels/scalar_compare.cc

```diff
@@ -439,6 +472,330 @@ struct ScalarMinMax {
+            num_valid++;
+          } else {
+            valid_cols[col] = string_view();
+          }
+        } else {
+          const auto& array = *batch[col].array();
+          if (!array.MayHaveNulls() ||
+              bit_util::GetBit(array.buffers[0]->data(), array.offset + row)) {
+            const auto data = array.GetValues<uint8_t>(1, /*absolute_offset=*/0);
+            valid_cols[col] = string_view(
+                reinterpret_cast<const char*>(data) + row * byte_width, byte_width);
+            num_valid++;
+          } else {
+            valid_cols[col] = string_view();
+          }
+        }
+      }
+
+      if (num_valid < batch.values.size() && !options.skip_nulls) {
+        // We had some nulls
+        builder.UnsafeAppendNull();
+        continue;
+      }
+      auto result = valid_cols.front();
+      for (size_t col = 1; col < batch.values.size(); ++col) {
+        const auto value = valid_cols[col];
+        if (value.empty()) {
+          DCHECK(options.skip_nulls);
+          continue;
+        }
+        result = result.empty() ? value : Op::Call(result, value);
+      }
+      if (result.empty()) {
+        builder.UnsafeAppendNull();
+      } else {
+        builder.UnsafeAppend(result);
+      }
+    }
+
+    std::shared_ptr<Array> string_array;
+    RETURN_NOT_OK(builder.Finish(&string_array));
+    *out = *string_array->data();
+    out->mutable_array()->type = batch[0].type();
+    DCHECK_EQ(batch.length, out->array()->length);
+    return Status::OK();
+  }
+
+  // Compute the length of the output for the given position, or -1 if it would be null.
+  static int64_t CalculateRowSize(const ElementWiseAggregateOptions& options,
+                                  const ExecBatch& batch, const int64_t index,
+                                  int32_t byte_width) {
+    const auto num_args = batch.values.size();
+    int32_t final_size = 0;
+    for (size_t i = 0; i < num_args; i++) {
+      bool valid = true;
+      if (batch[i].is_scalar()) {
+        const auto& scalar = *batch[i].scalar();
+        valid = scalar.is_valid;
+      } else {
+        const auto& array = *batch[i].array();
+        valid = !array.MayHaveNulls() ||
+                bit_util::GetBit(array.buffers[0]->data(), array.offset + index);
+      }
+      if (!valid) {
+        if (options.skip_nulls) {
+          continue;
+        }
+        return -1;
+      }
+      final_size = std::max(final_size, byte_width);
+    }
+    return final_size;
+  }
+};
+
+Result<ValueDescr> ResolveMinOrMaxOutputType(KernelContext*,
+                                             const std::vector<ValueDescr>& args) {
+  if (args.empty()) {
+    return null();
+  }
+  auto first_type = args[0].type;
+  for (size_t i = 1; i < args.size(); ++i) {
+    auto type = args[i].type;
+    if (*type != *first_type) {
+      return Status::NotImplemented(
+          "Different decimal types not implemented for {min, max}_element_wise");
```

Comment:
I think the error message is too specific, as you could have fixed-size binary types with different byte widths as well.